distsql: panic: sync: negative WaitGroup timer

I ran a 4-node test cluster with `photos` load and `COCKROACH_DISTSQL_MODE=on` as follows:

```
make test \
  PKG=./pkg/acceptance \
  TESTTIMEOUT=48h \
  TESTS=ContinuousLoad_Photos \
  TESTFLAGS='-v -remote -key-name google_compute_engine -cwd terraform -nodes 4 -tf.keep-cluster=failed -tf.cockroach-env="COCKROACH_DISTSQL_MODE=on, COCKROACH_DISTSQL_LOG_PLAN=1"'
```

After about two hours, this panic happened:

```
panic: sync: negative WaitGroup counter

goroutine 16535539 [running]:
panic(0x18213c0, 0xc427407c70)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
sync.(*WaitGroup).Add(0xc4240c2400, 0xffffffffffffffff)
        /usr/local/go/src/sync/waitgroup.go:75 +0x1fe
sync.(*WaitGroup).Done(0xc4240c2400)
        /usr/local/go/src/sync/waitgroup.go:100 +0x34
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*flowRegistry).finishInboundStreamLocked(0xc420241d90, 0xc422717920)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/flow_registry.go:267 +0x47
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*flowRegistry).FinishInboundStream(0xc420241d90, 0xc422717920)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/flow_registry.go:275 +0x6b
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).flowStreamInt(0xc420204000, 0x7f285ede3000, 0xc420a5acc0, 0x25fedc0, 0xc427407c00, 0x7f285ee5a220, 0xc425b21810)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:179 +0x400
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).FlowStream(0xc420204000, 0x25fedc0, 0xc427407c00, 0xc4206c0ca8, 0x20)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:185 +0xa5
github.com/cockroachdb/cockroach/pkg/sql/distsql._DistSQL_FlowStream_Handler(0x19c5b80, 0xc420204000, 0x25fb160, 0xc42541e900, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/api.pb.go:267 +0xbb
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processStreamingRPC(0xc4203a9360, 0x25ff300, 0xc420a1aab0, 0xc4275b24b0, 0xc4201f23c0, 0x25ca240, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:689 +0x702
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4203a9360, 0x25ff300, 0xc420a1aab0, 0xc4275b24b0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:773 +0xc36
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc420a30af0, 0xc4203a9360, 0x25ff300, 0xc420a1aab0, 0xc4275b24b0)
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:422 +0xab
created by github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:423 +0xa3
```

Logs for the crashed node:

[cockroach.stderr.zip](https://github.com/cockroachdb/cockroach/files/654990/cockroach.stderr.zip)

--- a/pkg/sql/distsql/flow.go
+++ b/pkg/sql/distsql/flow.go
@@ -342,13 +342,15 @@ func (f *Flow) Start(doneFn func()) {
 		f.Context, 1, "starting (%d processors, %d outboxes)", len(f.outboxes), len(f.processors),
 	)
 	f.status = FlowRunning
+
+	// Once we call RegisterFlow, the inbound streams become accessible; we must
+	// set up the WaitGroup counter before.
+	f.waitGroup.Add(len(f.inboundStreams) + len(f.outboxes) + len(f.processors))
+
 	f.flowRegistry.RegisterFlow(f.id, f, f.inboundStreams)
 	if log.V(1) {
 		log.Infof(f.Context, "registered flow %s", f.id.Short())
 	}
-	f.waitGroup.Add(len(f.inboundStreams))
-	f.waitGroup.Add(len(f.outboxes))
-	f.waitGroup.Add(len(f.processors))
 	for _, o := range f.outboxes {
 		o.start(&f.waitGroup)
 	}

